pkg/kstatus/observe/reader/caching_reader.go

```diff
@@ -0,0 +1,225 @@
+		mapping, err := mapper.RESTMapping(gk)
+		if err != nil {
+			return err
+		}
+		gvkNamespaceSet.add(gvkNamespace{
+			GVK:       mapping.GroupVersionKind,
+			Namespace: namespace,
+		})
+		genGKs, found := genGroupKinds[gk]
+		if found {
+			err := buildGvkNamespaceSet(mapper, genGKs, namespace, gvkNamespaceSet)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+type gvkNamespaceSet struct {
+	gvkNamespaces []gvkNamespace
+	seen          map[gvkNamespace]bool
+}
+
+func newGnSet() *gvkNamespaceSet {
+	return &gvkNamespaceSet{
+		gvkNamespaces: make([]gvkNamespace, 0),
+		seen:          make(map[gvkNamespace]bool),
+	}
+}
+
+func (g *gvkNamespaceSet) add(gn gvkNamespace) {
+	if _, found := g.seen[gn]; !found {
+		g.gvkNamespaces = append(g.gvkNamespaces, gn)
+		g.seen[gn] = true
+	}
+}
+
+// CachingClusterReader is an implementation of the ObserverReader interface that will
+// pre-fetch all resources needed before every sync loop. The resources needed are decided by
+// finding all combinations of GroupVersionKind and namespace referenced by the provided
+// identifiers. This list is then expanded to include any known generated resource types.
+type CachingClusterReader struct {
+	sync.RWMutex
+
+	// reader provides functions to read and list resources from the
+	// cluster.
+	reader client.Reader
+
+	// mapper is the client-side representation of the server-side scheme. It is used
+	// to resolve GroupVersionKind from GroupKind.
+	mapper meta.RESTMapper
+
+	// gns contains the slice of all the GVK and namespace combinations that
+	// should be included in the cache. This is computed based the resource identifiers
+	// passed in when the CachingClusterReader is created and augmented with other
+	// resource types needed to compute status (see genGroupKinds).
+	gns []gvkNamespace
+
+	// cache contains the resources found in the cluster for the given combination
+	// of GVK and namespace. Before each polling cycle, the framework will call the
+	// Sync function, which is responsible for repopulating the cache.
+	cache map[gvkNamespace]unstructured.UnstructuredList
+}
+
+// gvkNamespace contains information about a GroupVersionKind and a namespace.
+type gvkNamespace struct {
+	GVK       schema.GroupVersionKind
+	Namespace string
+}
+
+// Get looks up the resource identified by the key and the object GVK in the cache. If the needed combination
+// of GVK and namespace is not part of the cache, that is considered an error.
+func (c *CachingClusterReader) Get(_ context.Context, key client.ObjectKey, obj *unstructured.Unstructured) error {
+	c.RLock()
+	defer c.RUnlock()
+	gvk := obj.GetObjectKind().GroupVersionKind()
+	mapping, err := c.mapper.RESTMapping(gvk.GroupKind())
+	if err != nil {
+		return err
+	}
+	gn := gvkNamespace{
+		GVK:       gvk,
+		Namespace: key.Namespace,
+	}
+	cacheList, found := c.cache[gn]
+	if !found {
+		return fmt.Errorf("GVK %s and Namespace %s not found in cache", gvk.String(), gn.Namespace)
```

Comment:
How does this surface at the watcher level? i.e. does it just report errors on the affected resource(s)--errors that will likely go away in the next polling cycle--or is there some way this could end up being fatal to the watch?

Reply:
Currently any errors within a single observer are not considered fatal, so the error will be reported as part of an event for the particular resource. The observer polling will keep running.
Any errors encountered in the `Sync` function is considered fatal. This will cause the main observer loop to return an error event and then exit.
I'm not sure if this is the best way. Definitely that we could look more into.